crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -0,0 +1,221 @@
+use super::Command;
+use crate::world::World;
+
+/// # Safety
+///
+/// This function is only every called when the `command` bytes is the associated
+/// [`Commands`] `T` type. Also this only reads the data via `read_unaligned` so unaligned
+/// accesses are safe.
+unsafe fn invoke_command<T: Command>(command: *mut u8, world: &mut World) {
+    let command = if std::mem::size_of::<T>() > 0 {
+        command.cast::<T>().read_unaligned()
+    } else {
+        // NOTE: This is necessary because if the `CommandQueueInner` is only filled with 0-sized
+        // commands the `bytes` vec will never allocate. Then means that `bytes.as_ptr()` could be null
+        // and reading a null pointer is always UB.
+        // However according to https://doc.rust-lang.org/std/ptr/index.html
+        // "The canonical way to obtain a pointer that is valid for zero-sized accesses is NonNull::dangling"
+        // therefore the below code is safe to do.
+        let ptr = std::ptr::NonNull::<T>::dangling().as_ptr();
+        ptr.cast::<T>().read()
+    };
+    command.write(world);
+}
+
+struct CommandMeta {
+    offset: usize,
+    func: unsafe fn(value: *mut u8, world: &mut World),
+}
+
+#[derive(Default)]
+pub(crate) struct CommandQueueInner {
+    bytes: Vec<u8>,
+    metas: Vec<CommandMeta>,
+}
+
+// SAFE: All commands [`Command`] implement [`Send`]
+unsafe impl Send for CommandQueueInner {}
+
+// SAFE: `&CommandQueueInner` never gives access to the inner commands.
+unsafe impl Sync for CommandQueueInner {}
+
+impl CommandQueueInner {
+    /// Push a new `command` onto the queue.
+    #[inline]
+    pub fn push<C>(&mut self, command: C)
+    where
+        C: Command,
+    {
+        let size = std::mem::size_of::<C>();
+        let old_len = self.bytes.len();
+
+        self.metas.push(CommandMeta {
+            offset: old_len,
+            func: invoke_command::<C>,
+        });
+
+        if size > 0 {
+            self.bytes.reserve(size);
+
+            // SAFE: The internal `bytes` vector has enough storage for the
+            // command (see the call the `reserve` above), and the vector has
+            // its length set appropriately.
+            // Also `command` is forgotten at the end of this function so that
+            // when `apply` is called later, a double `drop` does not occur.
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    &command as *const C as *const u8,
+                    self.bytes.as_mut_ptr().add(old_len),
+                    size,
+                );
+                self.bytes.set_len(old_len + size);
+            }
+        }
+
+        std::mem::forget(command);
+    }
+
+    /// Invoke each command `func` for each inserted value with `world`
+    /// and then clears the internal bytes/metas command vectors.
+    pub fn apply(&mut self, world: &mut World) {
+        let mut bytes = std::mem::replace(&mut self.bytes, vec![]);
+        let byte_ptr = bytes.as_mut_ptr();
```

Comment:
This doesn't reuse the allocations, which is extremely unfortunate

A good enough solution would be to swap bytes back in at the end of the function, after clearing it.

Reply:
Another option would be to use `unsafe { self.bytes.set_len(0); }` instead of the replace. This is using raw pointers anyway.

Reply:
Yep, I like this more.

Reply:
Something else I just considered. 
If a command panics and you do some `catch_unwind` stuff. 
The rest of the commands in the queue won't be dropped and could possibly leak memory/not clean up resources.

How do we feel about this?

An option is to also store a `drop` `fn` and add a `impl Drop` for `CommandQueueInner`.
@DJMcNab @bjorn3 thoughts?

Reply:
Leaking on panic is absolutely fine.

Reply:
Just update the implementation to manually `.set_len(0)`.